src/LanguageServer/ProtocolUnitTests/LspServicesTests.cs

```diff
@@ -0,0 +1,120 @@
+        await using var server = await CreateTestLspServerAsync("", mutatingLspWorkspace, initializationOptions: new() { ServerKind = WellKnownLspServerKinds.CSharpVisualBasicLspServer }, composition);
+
+        var lspService = server.GetRequiredLspService<TestLspService>();
+        Assert.True(lspService is CSharpLspService);
+
+        var lspServiceFromFactory = server.GetRequiredLspService<TestLspServiceFromFactory>();
+        Assert.Equal(typeof(CSharpLspServiceFactory).Name, lspServiceFromFactory.FactoryName);
+    }
+
+    [Theory, CombinatorialData]
+    public async Task ReturnsAnyLspService(bool mutatingLspWorkspace)
+    {
+        var composition = base.Composition.AddParts(typeof(AnyLspService), typeof(AnyLspServiceFactory));
+        await using var server = await CreateTestLspServerAsync("", mutatingLspWorkspace, initializationOptions: new() { ServerKind = WellKnownLspServerKinds.CSharpVisualBasicLspServer }, composition);
+
+        var lspService = server.GetRequiredLspService<TestLspService>();
+        Assert.True(lspService is AnyLspService);
+
+        var lspServiceFromFactory = server.GetRequiredLspService<TestLspServiceFromFactory>();
+        Assert.Equal(typeof(AnyLspServiceFactory).Name, lspServiceFromFactory.FactoryName);
+    }
+
+    [Theory, CombinatorialData]
+    public async Task DuplicateSpecificServicesThrow(bool mutatingLspWorkspace)
+    {
+        var composition = base.Composition.AddParts(typeof(CSharpLspService), typeof(CSharpLspServiceFactory), typeof(DuplicateCSharpLspService), typeof(DuplicateCSharpLspServiceFactory));
+        await Assert.ThrowsAnyAsync<Exception>(async () => await CreateTestLspServerAsync("", mutatingLspWorkspace, initializationOptions: new() { ServerKind = WellKnownLspServerKinds.CSharpVisualBasicLspServer }, composition));
+    }
+
+    [Theory, CombinatorialData]
+    public async Task DuplicateAnyServicesThrow(bool mutatingLspWorkspace)
+    {
+        var composition = base.Composition.AddParts(typeof(AnyLspService), typeof(AnyLspServiceFactory), typeof(DuplicateAnyLspService), typeof(DuplicateAnyLspServiceFactory));
+        await Assert.ThrowsAnyAsync<Exception>(async () => await CreateTestLspServerAsync("", mutatingLspWorkspace, initializationOptions: new() { ServerKind = WellKnownLspServerKinds.CSharpVisualBasicLspServer }, composition));
+    }
+
+    internal class TestLspService : ILspService { }
+
+    internal record class TestLspServiceFromFactory(string FactoryName) : ILspService { }
+
+    internal class TestLspServiceFactory : ILspServiceFactory
+    {
+        public ILspService CreateILspService(LspServices lspServices, WellKnownLspServerKinds serverKind) => new TestLspServiceFromFactory(this.GetType().Name);
+    }
+
+    [ExportStatelessLspService(typeof(TestLspService), ProtocolConstants.RoslynLspLanguagesContract, WellKnownLspServerKinds.CSharpVisualBasicLspServer), Shared]
```

Comment:
>WellKnownLspServerKinds.CSharpVisualBasicLspServer [](http://example.com/codeflow?start=101&length=50)

Might be nice to have another language also exported for these

Reply:
Didn't quite get what you mean

Reply:
Sorry, poorly phrased. I had thought it might be useful to specify a different WellKnownLspServerKinds (eg: XamlLspServer) to verify we weren't getting back those services from the GetRequiredLspService calls. But, maybe that's covered elsewhere.

Reply:
I feel like I wrote a test for this at some point, but I can't find it now.  Adding one.